### DISCUSSION TITLE
Setting up Windows 10 network connection, for device network adapter via USB (RNDIS)?

### DISCUSSION THREAD

**Discussion by sdbbs** [hathach/tinyusb#1550] [Q&A]
I am woring with an RP2040 Pico board, and I wanted to explore the possibilities for the Pico to present itself as an Ethernet network adapter to the PC, when it is connected to the PC over USB cable.

The only full buildable example for this platform which uses tinyusb I could find is https://github.com/maxnet/pico-webserver ; I can compile and flash this program on the Pico fine, no problem there.

The pico-webserver program should serve a webpage on IP address 192.168.7.1:

> Runs a webinterface at `http://192.168.7.1/`

Note that the Windows 10 PC that I'm trying this on already has an actual Ethernet connection to a wired LAN, whose DHCP server assigns addresses in 10.0.* IP address range.

So, after flashing pico-webserver, I do get a network adapter device in Windows 10, which is great - the problem is, I cannot access it.

I played around with the settings a bit such as trying to force a static IP, but it did not work; so I tried to restore the defaults, not sure if I managed to do so fully - but this is what I can see in Windows 10 related settings:

go to Control Panel\All Control Panel Items\Network and Sharing Center

![image](https://user-images.githubusercontent.com/48317456/178140967-7c35f3c1-4a7e-44a9-809a-107ed7f76f7c.png)

In my case, the network "connector" that the pico-webserver instantiated is "Ethernet 6"; I don't mind that it is reported as "unidentified network" - but I am not sure why do all of those *other* network interfaces show up (might be because they are "virtual", created by software like VMWare or VirtualBox?)

Note that if I click "Ethernet 6" here, I get status:

![image](https://user-images.githubusercontent.com/48317456/178143756-5170f6ae-7642-4c3e-9d97-c787b785e834.png)

... where I can note, that there are both "Sent" and "Received" bytes; so on some level, the Windows 10 PC, and the pico-webserver device, do talk to each other - which is good.

Click "change adapter settings" -> Control Panel\Network and Internet\Network Connections

![image](https://user-images.githubusercontent.com/48317456/178141161-6151c2bf-9da2-4a8c-9cd0-e3abeca7a603.png)

Here we can see that "Ethernet 6" is "Remote NDIS based Internet Sharing"; I don't have much experience with this, and as per https://en.wikipedia.org/wiki/Ethernet_over_USB I sort of expected the pico-webserver network device would have been a CDC class device (CDC-ECM, CDC-EEM or CDC-NCM), but this confirms that it is RNDIS.

Right-click on "Ethernet 6", click Properties (took forever to montage this screenshot manually):

![image](https://user-images.githubusercontent.com/48317456/178142064-a841b166-a5ae-4209-87c7-b2acf72b9cb8.png)

We can see a whole bunch of drivers hooked into this device.

Typically, I would have set static IP by double-clicking the "Internet Protocol Version 4 (TCP/IPv4)" node here, which by default looks like this:

![image](https://user-images.githubusercontent.com/48317456/178142302-23e6522a-5a98-41bc-bce0-9cbf6e96d58b.png)

I've tried both with setting "Use the following IP address" with IP address 192.168.7.1, subnet mask 255.255.255.0, default gateway 192.168.7.1 - and with "Obtain an IP address automatically" but with "Alternate Configuration" set to "User configured" and the same settings - and I cannot get the static IP to work. 
I've additionally tried the same, with "VMWare Bridge Protocol", "VirtualBox NDIS6 Bridged Networking Driver", "Bridge Driver" and  "Internet Protocol Version 6 (TCP/IPv6)" drivers unchecked from "Ethernet 6" device - still cannot get the static IP to work.

Here is what Device Manager (`devmgmt.msc`) sees, including the screens for right-click/Properties on the "Remote NDIS based Internet Sharing Device":

![image](https://user-images.githubusercontent.com/48317456/178143090-0de7d141-34cc-4bc1-874b-82d5bdd851d9.png)

We can see USB VID:PID of the device is CAFE:4020, and that Windows 10 loaded `usbrndis6` driver for it.

The command `ipconfig /all` reports for this network interface:

```none
C:\WINDOWS\system32> ipconfig /all

Windows IP Configuration
...
Ethernet adapter Ethernet 6:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Remote NDIS based Internet Sharing Device
   Physical Address. . . . . . . . . : 02-02-84-6A-96-00
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::69f4:f173:81b2:91d9%133(Preferred)
   Autoconfiguration IPv4 Address. . : 169.254.145.217(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.0.0
   Default Gateway . . . . . . . . . :
   DHCPv6 IAID . . . . . . . . . . . : -2063465852
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-21-CC-8A-2E-6C-4B-90-47-A6-80
   DNS Servers . . . . . . . . . . . : fec0:0:0:ffff::1%1
                                       fec0:0:0:ffff::2%1
                                       fec0:0:0:ffff::3%1
   NetBIOS over Tcpip. . . . . . . . : Enabled
...

C:\WINDOWS\system32>ping 169.254.145.217

Pinging 169.254.145.217 with 32 bytes of data:
Reply from 169.254.145.217: bytes=32 time<1ms TTL=128
Reply from 169.254.145.217: bytes=32 time<1ms TTL=128
Reply from 169.254.145.217: bytes=32 time<1ms TTL=128
Reply from 169.254.145.217: bytes=32 time<1ms TTL=128

Ping statistics for 169.254.145.217:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 0ms, Average = 0ms

C:\WINDOWS\system32>ping 192.168.7.1

Pinging 192.168.7.1 with 32 bytes of data:
Request timed out.
Request timed out.
Request timed out.
Request timed out.

Ping statistics for 192.168.7.1:
    Packets: Sent = 4, Received = 0, Lost = 4 (100% loss),
```

So, somehow the IP address ended up being 169.254.145.217 (prob. due to "autoconfiguration");  this address 169.254.145.217  is actually pingable - but 192.168.7.1 is not; I tried reading from both these IP addresses (using `wget` in MINGW64 `bash` shell), and I cannot connect in either case:

```none
$ wget 192.168.7.1 -O-
--2022-07-10 13:40:53--  http://192.168.7.1/
Connecting to 192.168.7.1:80... failed: Connection timed out.
Retrying.

--2022-07-10 13:41:15--  (try: 2)  http://192.168.7.1/
Connecting to 192.168.7.1:80...
^C

$ wget 169.254.145.217 -O-
--2022-07-10 14:00:09--  http://169.254.145.217/
Connecting to 169.254.145.217:80... failed: Connection refused.
```

And finally, this is what the command line Device Manager, `devcon.exe`, sees about this VID:

```none
$ /c/bin/devcon.exe driverfiles '*CAFE*'
USB\VID_CAFE&PID_4020\E660D4A0A75EA238
    Name: Remote NDIS based Internet Sharing Device
    Driver installed from C:\WINDOWS\INF\wceisvista.inf [RNDIS.NT.5.1]. 2 file(s) used by driver:
        C:\WINDOWS\system32\DRIVERS\rndismp6.sys
        C:\WINDOWS\system32\DRIVERS\usb80236.sys
1 matching device(s) found.
```

So: how can I set up this RNDIS network adapter in Windows 10, so I can read webpages from the pico-webserver device on this OS (which is by default set to IP 192.168.7.1, and port 80 for HTTP)?

**Comment by pigrew** [comment]
For whatever reason, it looks like DHCP wasn't responding.

The first thing I'd try is to statically assign the interface an IP of 192.168.7.5, and then see if I can connect to 192.168.7.1. Someone more familiar with the RNDIS port might be able to comment on why DHCP is missing.

I'd also suggest directly trying the [net_lwip_webserver](https://github.com/hathach/tinyusb/tree/master/examples/device/net_lwip_webserver) example in TinyUSB, as I believe there have been many fixes since that other example was created.

**Comment by sdbbs** [comment]
Many thanks for the response, @pigrew : 

> I'd also suggest directly trying the [net_lwip_webserver](https://github.com/hathach/tinyusb/tree/master/examples/device/net_lwip_webserver) example in TinyUSB, as I believe there have been many fixes since that other example was created.

Good to know, was not aware of that - however, I did manage to get some results with the old code referred in OP, so what I'm reporting here is in respect to that

> For whatever reason, it looks like DHCP wasn't responding.

Thinking about this, I realized I'm missing a model of how this should work. What is the DHCP client: does the MCU code (tinyusb from within the pico-webserver app) issue a DHCP request - or does Windows do that on behalf of the network device? Also, what should be the DHCP server: should I expect Windows 10 to overtake that role (which is what, I guess, happens when we have "autoconfiguration" of IP addresses) - or should it be the DHCP server on the router that manages the LAN the Windows 10 PC is connected on?

EDIT: just realized [webserver.c](https://github.com/maxnet/pico-webserver/blob/main/webserver.c) calls a function defined in [tusb_lwip_glue.c](https://github.com/maxnet/pico-webserver/blob/main/tusb_lwip_glue.c) called `dhcpd_init`, so it uses DHCP - although the name is somewhat confusing to me, since "dhcpd" for me implies "dhcp daemon" - that is, server - and I find the MCU taking up the role of DHCP server kind of strange ...

In line with this, I noticed in [tusb_lwip_glue.c](https://github.com/maxnet/pico-webserver/blob/main/tusb_lwip_glue.c):

```c
/* this is used by this code, ./class/net/net_driver.c, and usb_descriptors.c */
/* ideally speaking, this should be generated from the hardware's unique ID (if available) */
/* it is suggested that the first byte is 0x02 to indicate a link-local address */
const uint8_t tud_network_mac_address[6] = {0x02,0x02,0x84,0x6A,0x96,0x00};
```

I was not aware that MAC address could "indicate a link-local address", but I found this https://ma.ttias.be/stateless-autoconfiguratio-replace-dhcp-ipv6/ : 

> MAC Address Layout
>
> The first part of the MAC address is guaranteed to be unique by IEEE, the second part is guaranteed to be unique by the owner the of OUI (=the company which was assigned that specific OUI; ie: Apple, Xerox, HP, …).
>
> The MAC address, however, has a so called “universal / local bit”. A specific bit (0/1) to indicate whether the MAC address is globally unique (provided by the hardware supplier), or not (it might be altered afterwards) and whether the MAC address is a multicast address, or a unicast address.
>
> If a MAC address is not available, the “universal / local bit” is set to one, to indicate that the MAC address isn’t globally unique, and can’t be used universally. The end result – a combination of the 64 bit router-supplied prefix, and the MAC address – will form a Modified EUI-64 instead of a regular EUI-64 (because the “u/l bit” was flipped).

I'm not sure if this is related to the "indicate a link-local address" comment above - but, maybe it is the specific MAC address that the MCU code uses, that makes the IP address setup on the PC difficult?

> The first thing I'd try is to statically assign the interface an IP of 192.168.7.5, and then see if I can connect to 192.168.7.1. 

Thank you, that was it! I noticed in Device Manager/Network Adapters/Remote NDIS based Internet Sharing Device/Advanced there is a field, "Network Address".

I have tried both with setting this field, vs keeping it "not present" - there seems to be not much difference.

With "Network Address" set to 192.168.7.5:

![ethernet_6_01_devnet_set](https://user-images.githubusercontent.com/48317456/178397715-9d9e9566-3e58-40af-885c-f2a0c09f91bd.png)

* Setting Control Panel/All Control Panel Items/Network Connections/Ethernet 6 Properties/Networking tab/Internet Protocol Version 4 (TCP/IPv4) to "Obtain an IP address automatically" (dynamic IP) does not work (device gets Autoconfiguration IPv4 Address. . : 169.254.145.217)
* Setting Control Panel/All Control Panel Items/Network Connections/Ethernet 6 Properties/Networking tab/Internet Protocol Version 4 (TCP/IPv4) to "Use the following IP address"  (static IP) works for ip address 192.168.7.5 and subnet mask 255.255.255.0; note:
    * In this case, Windows 10 refuses to accept a "default gateway" setting
    * For some reason, Windows 10 shows this network adapter as greyed out/Disabled - and it cannot be enabled either - however both 192.168.7.5 and 192.168.7.1 are pingable, and reading the webpage works

In this case

```none
ipconfig /all:
...
Ethernet adapter Ethernet 6:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Remote NDIS based Internet Sharing Device
   Physical Address. . . . . . . . . : 02-02-84-6A-96-00
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   IPv4 Address. . . . . . . . . . . : 192.168.7.5(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . : 0.0.0.0
   NetBIOS over Tcpip. . . . . . . . : Enabled
...
```

--------

With "Network Address" unset ("not present"):
 
![ethernet_6_02_devnet_unset](https://user-images.githubusercontent.com/48317456/178398773-aa32d4b4-0267-443d-b4e7-b4b98702a51a.png)

* Setting Control Panel/All Control Panel Items/Network Connections/Ethernet 6 Properties/Networking tab/Internet Protocol Version 4 (TCP/IPv4) to "Obtain an IP address automatically" (dynamic IP) does not work (device again gets Autoconfiguration IPv4 Address. . : 169.254.145.217)
* Setting Control Panel/All Control Panel Items/Network Connections/Ethernet 6 Properties/Networking tab/Internet Protocol Version 4 (TCP/IPv4) to "Use the following IP address"  (static IP) works for ip address 192.168.7.5 and subnet mask 255.255.255.0; note:
    * In this case, Windows 10 again refuses to accept a "default gateway" setting
    * However, this time the network adapter in Windows 10 is shown as Enabled

ipconfig reports the same as in the previous case; also noticed here that:

```
C:\WINDOWS\system32>ipconfig /renew "Ethernet 6"

Windows IP Configuration

Adapter Ethernet 6 is not enabled for DHCP.
```

--------

Ultimately, what I want is to come to the point, where upon plugging the device via USB, the corresponding interface automatically gets assigned the right IP address (say, maybe I could "hint" from the MCU program, that the device would want any address in the 192.168.7.* range apart from 192.168.7.1) - is this in principle possible, and if so, what should I look into next, to get something like that working?

EDIT2: I guess that is how it should work, since [tusb_lwip_glue.c](https://github.com/maxnet/pico-webserver/blob/main/tusb_lwip_glue.c) uses `dhcpd_init()` -> `dhserv_init(&dhcp_config)` where struct `dhcp_config` has a field `entries`, which is:

```c
/* database IP addresses that can be offered to the host; this must be in RAM to store assigned MAC addresses */
static dhcp_entry_t entries[] =
{
    /* mac ip address                          lease time */
    { {0}, IPADDR4_INIT_BYTES(192, 168, 7, 2), 24 * 60 * 60 },
    { {0}, IPADDR4_INIT_BYTES(192, 168, 7, 3), 24 * 60 * 60 },
    { {0}, IPADDR4_INIT_BYTES(192, 168, 7, 4), 24 * 60 * 60 },
};
```

... so I guess there is a glitch somewhere, in propagating this information to the host Windows 10 PC ...

**Reply by pigrew** [reply]
The USB device is the "DHCP server", and a few assignable IP addresses (192.168.7.2 through 192.168.7.4) are hardcoded. Windows in this case is the client. Windows should be requesting an IP address from the USB device. For whatever reason, Windows is unable to request an IP address from the USB device. I'm not sure where to go from here (I have not followed the development of the TinyUSB networking class).

My guess is that there was some bug that was fixed in the networking code at some point, but I'm not sure.... You might be able to enable/configure the serial debug output from TinyUSB and check if it's reporting any sort of errors.

**Comment by sdbbs** [comment]
Many thanks for the response @pigrew : 

> I'd also suggest directly trying the [net_lwip_webserver](https://github.com/hathach/tinyusb/tree/master/examples/device/net_lwip_webserver) example in TinyUSB, as I believe there have been many fixes since that other example was created.

After some trouble ( https://github.com/raspberrypi/pico-examples/issues/238 ), I finally managed to build and flash the `net_lwip_webserver` example - and I can finally report success (see below)!

> The USB device is the "DHCP server", and a few assignable IP addresses (192.168.7.2 through 192.168.7.4) are hardcoded. Windows in this case is the client. Windows should be requesting an IP address from the USB device.

Thanks for confirming this; the device being the DHCP server really did not fit my mental model of how DHCP works (then again, I'm not a networking expert), so it was really good to see a confirmation that my reading of the code was correct after all. And now that I've seen it in action, I think I understand better what is going on.

Note that I've used `tinyusb` where it is included as submodule of pico-sdk, and I've had revision 4bfab30c (Oct 19, 2021) of tinyusb there; and `lwip` as a submodule of `tinyusb` there is at commit 159e31b68.

Anyways, after building and flashing this example, I got a "Remote NDIS based Internet Sharing Device #2" in Device Manager; by default it has Network Address: Not Present - and interestingly:

![image](https://user-images.githubusercontent.com/48317456/178495911-a586d68e-8773-4afa-abc7-dca8fb5a9742.png)

... the Events tab shows only two events - whereas the pico-webserver examples had at least six (on [this image](https://user-images.githubusercontent.com/48317456/178143090-0de7d141-34cc-4bc1-874b-82d5bdd851d9.png))

This ends up as "Ethernet 7" adapter in Network Connections, and `ipconfig /all` now reports for it:

```
Ethernet adapter Ethernet 7:

   Connection-specific DNS Suffix  . : usb
   Description . . . . . . . . . . . : Remote NDIS based Internet Sharing Device #2
   Physical Address. . . . . . . . . : 02-02-84-6A-96-00
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::348c:efd4:ee1:b66c%132(Preferred)
   IPv4 Address. . . . . . . . . . . : 192.168.7.2(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Lease Obtained. . . . . . . . . . : Tuesday, July 12, 2022 2:32:20 PM
   Lease Expires . . . . . . . . . . : Wednesday, July 13, 2022 2:32:19 PM
   Default Gateway . . . . . . . . . :
   DHCP Server . . . . . . . . . . . : 192.168.7.1
   DHCPv6 IAID . . . . . . . . . . . : -2080243068
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-21-CC-8A-2E-6C-4B-90-47-A6-80
   DNS Servers . . . . . . . . . . . : 192.168.7.1
   NetBIOS over Tcpip. . . . . . . . : Enabled
```

Yup, so the interface got 192.168.7.2 - one of the hardcoded DHCP addresses, and so I could read `http://192.168.7.1` immediately after flashing, without having to do anything extra (exactly as I imagined it should work)!

So, problem solved - and thanks a lot for the assistance!

**Comment by hathach** [comment]
@sdbbs if problem is solved, please select one of the response and "Mark as answer". This will help others with similar issue. Thanks
